### PR TITLE
Fixes and improvements in failsafe

### PR DESCRIPTION
**Pull request by CyberDem0n**
1. Fix problem with logical slots not advancing when only the primary lost access to DCS
2. Don't let Patroni to join as a raft voting member when running failsafe behave tests. It allows to test exactly the same conditions as for other DCS
3. Speed up dcs_failsafe_mode behave tests by getting rid from long sleeps, slight reshuffling of places when we start/stop outage, and by killing Patroni/Postgres to avoid long shutdown due to the leader key removal attempts.

### CODE REVIEW

File: features/environment.py [hunk 1]
```diff
@@ -184,7 +184,7 @@ def _make_patroni_test_config(self, name, custom_config):
             config.pop('etcd', None)
 
         raft_port = os.environ.get('RAFT_PORT')
-        if raft_port:
+        if raft_port and not self._output_dir.endswith('dcs_failsafe_mode'):
```
Comment (hughcapet) [inline features/environment.py:187]: Can we have the point from the PR desc as a comment here as well?

File: patroni/ha.py [hunk 2]
```diff
@@ -1695,15 +1692,18 @@ def _handle_dcs_error(self):
...
-                if cluster:
-                    self.state_handler.slots_handler.sync_replication_slots(cluster,
-                                                                            self.patroni.nofailover,
-                                                                            self.patroni.replicatefrom,
-                                                                            self.is_paused())
-
+                self._sync_replication_slots()
         return 'DCS is not accessible'
 
+    def _sync_replication_slots(self):
```
Comment (hughcapet) [inline patroni/ha.py:1698]: and the explanation how syncing of replication slots (=this function) works both for failsafe mode enabled and failing dcs (incl. this corner case of readonly dcs) and in the "normal operational mode"? (now, after an hour of despair, I think I got it)

File: patroni/ha.py [hunk 3]
```diff
@@ -1694,15 +1694,33 @@ def _handle_dcs_error(self):
...
+        cluster = self._failsafe.update_cluster(self.cluster)\
+            if self.is_failsafe_mode() and not self.is_leader() else self.cluster
+        if cluster:
+            slots = self.state_handler.slots_handler.sync_replication_slots(cluster,
                                                                             self.patroni.nofailover,
                                                                             self.patroni.replicatefrom,
                                                                             self.is_paused())
-
-        return 'DCS is not accessible'
+            # Don't copy replication slots if failsafe_mode is is active
```
Comment (hughcapet) [inline patroni/ha.py:1722]:
```suggestion
            # Don't copy replication slots if failsafe_mode is active
```

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Pull request merged**